packages/layers/src/scale-bar-layer.js

```diff
@@ -76,15 +77,39 @@ const ScaleBarLayer = class extends CompositeLayer {
       2 ** (-zoom + 1.5),
       (boundingBox[2][1] - boundingBox[0][1]) * 0.007
     );
-    const numUnits = barLength * size;
+
+    // Initialize values for the non-snapped case.
+    let adjustedBarLength = barLength;
+    let displayNumber = (barLength * size).toPrecision(5);
+    let displayUnit = unit;
+    if (snap) {
+      // Convert `size` to meters, since `snapValue`
+      // assumes the value is in meters.
+      const meterSize = sizeToMeters(size, unit);
+      const numUnits = barLength * meterSize;
+      // Get snapped value in original units and new units.
+      const [snappedOrigUnits, snappedNewUnits, snappedUnitPrefix] =
+        snapValue(numUnits);
+      // We adjust the bar length by using the ratio of the snapped
+      // value in original units to the original value in original units.
+      adjustedBarLength =
+        (numUnits * (snappedOrigUnits / numUnits)) / meterSize;
```

Comment:
```suggestion
      // We adjust the bar length by using the ratio of the snapped
      // value in original units to the original value passed to `snapValue` (which is based on `meterSize`).
      adjustedBarLength = snappedOrigUnits / meterSize;
```

I think this is simpler and also works?